sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,247 @@
+                          "0.".format(", ".join(str(i) for i in same_min_max)))
+
+        with np.errstate(divide='ignore', invalid='ignore'):
+            bin_widths = ptp / self.n_bins_
+        bin_widths[ignored] = 0
+        self.bin_width_ = bin_widths
+        return self
+
+    @staticmethod
+    def _check_n_bins(n_bins, n_features, ignored):
+        if isinstance(n_bins, numbers.Number):
+            if n_bins < 2:
+                raise ValueError("Discretizer received an invalid number "
+                                 "of bins. Received {0}, expected at least 2."
+                                 .format(n_bins))
+            return np.ones(n_features) * n_bins
+
+        n_bins = check_array(n_bins, dtype=int, copy=True, ensure_2d=False)
+
+        if n_bins.ndim > 1 or n_bins.shape[0] != n_features:
+            raise ValueError("n_bins must be a scalar or array "
+                             "of shape (n_features,).")
+
+        bad_nbins_value = n_bins < 2
+        is_ignored = np.zeros(n_features).astype(bool)
+        is_ignored[ignored] = True
+
+        violating_indices = np.where(~is_ignored & bad_nbins_value)[0]
+        if violating_indices.shape[0] > 0:
+            indices = ", ".join(str(i) for i in violating_indices)
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins at indices {}. Number of bins "
+                             "must be at least 2."
+                             .format(indices))
+        return n_bins
+
+    @staticmethod
+    def _check_ignored_features(ignored, n_features):
+        if ignored is None:
+            return np.array([], dtype='int64')
+
+        ignored = check_array(ignored, ensure_2d=False, dtype=int)
+        ignored = column_or_1d(ignored)
+
+        if len(set(ignored)) != ignored.shape[0]:
+            raise ValueError("Duplicate ignored column indices found.")
+
+        if np.all(ignored >= 0) and np.all(ignored < n_features):
+            return ignored
+
+        raise ValueError("Invalid ignored feature index.")
+
+    def transform(self, X):
+        """Discretizes the data.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            Data to be discretized.
+
+        Returns
+        -------
+        Xt : numeric array-like, shape (n_samples, n_features)
+            Data in the binned space.
+        """
+        X = self._check_X_post_fit(X)
+
+        return _transform_selected(X, self._transform,
+                                   self.transformed_features_, copy=True,
+                                   retain_order=True)
+
+    def _transform(self, X):
+        """Performs transformation on X, with no ignored features."""
+        trans = self.transformed_features_
+
+        X -= self.offset_[trans]
+        with np.errstate(divide='ignore', invalid='ignore'):
+            X //= self.bin_width_[trans]
```

Comment:
I think it was this kind of test I had issues with, running on a similar implementation:
```
@pytest.mark.parametrize(['n_bins', 'Xt_expected'],
                         [(2, [0, 0, 1, 1, 1]),
                          (3, [0, 0, 1, 2, 2]),
                          (4, [0, 1, 2, 3, 3]),
                          (5, [0, 1, 2, 3, 4]),
                          (6, [0, 1, 3, 4, 5]),
                          ])
@pytest.mark.parametrize('X,min_,ptp_',
                         [([[.2], [.4], [.6], [.8], [1.0]], .2, .8),
                          ([[2], [4], [6], [8], [10]], 2, 8),
                          ])
def test_discretizer_n_bins(n_bins, X, Xt_expected, min_, ptp_):
    est = Discretizer(n_bins=n_bins).fit(X)
    assert_array_equal(np.reshape(Xt_expected, (-1, 1)), est.transform(X))
    # check out of training range
    assert_array_equal([[0], [n_bins - 1]], est.transform([[0.1], [100]]))
    assert_array_equal(min_, est.min_)
    assert_array_equal(ptp_, est.ptp_)
```